### ISSUE TITLE
Only plain literals should be parsed

### ISSUE THREAD

**Issue by mbarkhau** [google/python-fire#97]
@dbieber correctly pointed out in #95 that there is a more general problem with the parsing code. This is a follow up issue to that issue which only dealt with the most common case of a single argument.

Here are some cases demonstrating
```bash
$ python3 -c "import fire.parser as p; v = p.DefaultParseValue('[1+1]'); print(type(v), v)"
<class 'list'> [2]
$ python3 -c "import fire.parser as p; v = p.DefaultParseValue('{key: 1+1}'); print(type(v), v)"
<class 'dict'> {'key': 2} 
```

Some options are:
1. Accept the existing behaviour.
2. Always parse the whole argument as a string by doing `raise ValueError` in ``fire.parser._LiteralEval`
3. Parse only the sub expressions as strings

For 2. and 3. I think these would be the valid types from the `ast` module. 

 - `Str`
 - `Name` (converted to `Str`)
 - `Num`
 - `List`
 - `Tuple`
 - `Dict`
 - `Set`
 - `UnaryOp`

**Comment by dbieber** [comment]
First, something that confuses me:
The [literal_eval documentation](https://docs.python.org/2/library/ast.html#ast.literal_eval) says it is not capable of evaluating arbitrarily complex expressions, **for example involving operators** or indexing.

This made me wonder, why does ast.literal_eval("1+1") evaluate to 2? A: It doesn't.
So why does DefaultParseValue('[1+1]') give [2]? A: It doesn't.
Turns out this is a Python 2 vs Python 3 distinction. ast.literal_evak("1+1") does give 2 in Python 3 :(.

So if we can get a port of Python 2's literal_eval into Python 3, I think that resolves our issues and we can remove the BinOp special case altogether. I haven't investigated whether this exists yet.

--------
Aside:
Something to keep in the back of your mind as you think about this: it would be _amazing_ if there were a standard way of parsing literals from the command line that was consistent across languages. This way _if_ someone were to implement, say, Java Fire, _and_ if someone were to implement piping args from the command line into Fire CLIs, then Fire could become a mechanism for interoperating across languages.

**Comment by dbieber** [comment]
The Python 3 documentation for literal_eval also suggests it won't parse expressions with operators.
The discussion at https://bugs.python.org/issue22525 suggests the reason that + and - are supported is to support _complex_ literals like 4+2i. I don't think that's sufficient justification for considering "1+1" a literal; from my understanding, "1+1" is not a literal and so I'm inclined to think this is a Python 3 bug.

**Comment by dbieber** [comment]
Opened issue in Python bug tracker at https://bugs.python.org/issue31778

**Comment by dbieber** [comment]
Even if this issue with Python is fixed, we'll still want Fire to work with builds of Python that have the current behavior. So, we should walk the AST and check for BinOps ourself, and raise a ValueError if there is a BinOp in our AST.

The ast.literal_eval code ([Python 2.7.3](https://github.com/enthought/Python-2.7.3/blob/69fe0ffd2d85b4002cacae1f28ef2eb0f25e16ae/Lib/ast.py#L67) and [Python 3](https://github.com/python/cpython/blob/ef611c96eab0ab667ebb43fdf429b319f6d99890/Lib/ast.py#L76)) may be useful for reference.

**Comment by dbieber** [comment]
The next steps here are to check for BinOps in the ast of the input. This will happen here: 

https://github.com/google/python-fire/blob/72604f40314008e562ba47936dcc183b51166b72/fire/parser.py#L97-L98

If there are any BinOps in the ast (this requires traversing the ast to determine), then we will raise a ValueError as above, rather than parsing this non-literal. The only exception to this is that the BinOps + and - used in the context of expressing a complex number (e.g. 2 + 3j) are allowed.

The [Python 2.7.3 implementation of literal_eval](https://github.com/enthought/Python-2.7.3/blob/69fe0ffd2d85b4002cacae1f28ef2eb0f25e16ae/Lib/ast.py#L67) demonstrates how we can do this AST walk to determine if any non-allowed BinOps are present.

**Comment by alexshadley** [comment]
Looks like I've got some spare time in the next week, I'll give this one a whirl.

**Comment by dbieber** [comment]
Hey Alex, keep us posted if you try anything.

**Comment by bcluyse** [comment]
Hi, can I have a look at this issue?

**Comment by alexshadley** [comment]
Yes!  Sorry for the long radio silence, I should've commented earlier that I was no longer working on this one.
